src/python/pants/backend/shell/dependency_inference.py

```diff
@@ -115,6 +115,9 @@ async def parse_shell_imports(
             input_digest=input_digest,
             description=f"Detect Shell imports for {request.fp}",
             level=LogLevel.DEBUG,
+            # We expect this to always fail, but it should still be cached because the process is
```

Comment:
I think the final clause is irrelevant in general. We expect all processes are deterministic by design. Perhaps we made a dubious affordance for flaky tests and flaky networks in our core handling though; thus forcing you to state what should be the status quo here?

Reply:
Yeah this PR made me start to wonder if we should be caching linter failures, for example...?

I think it makes sense for tests because those run user code and often can flake. 

For linters, formatters, and type checkers, those should not flake. And they also shouldn't be making network requests. But I'm not sure? There's less benefit to caching those because nothing else depends on those results, whereas here, not caching majorly slows down everything because dep inference is not memoized.

Reply:
+1. I've felt for a while that the default should be ALWAYS.

Reply:
That makes sense to me.

Reply:
I started to do this as a quick thing to knock off the todo list, but it occurred to me that infrastructure errors like OOM failures (at the JVM or kernel level) and etc should likely not be cached. It's hard to change the default here without a deeper semantic understanding of whether a failure is infrastructure-level or deterministic.

I could imagine adding that sort of semantic filter to `Process`... something like `Process(.., result_classifier=..)`, but without that, changing the default would likely result in some incorrectly cached infra errors.

Reply:
Yeah, we should be very careful not to cache transient errors...